dockstore-webservice/src/main/java/io/dockstore/webservice/resources/TokenResource.java

```diff
@@ -588,6 +595,82 @@ public Token addBitbucketToken(@ApiParam(hidden = true) @Auth User user, @QueryP
         }
     }
 
+
+
+    @GET
+    @Timed
+    @UnitOfWork
+    @Path("/zenodo.org")
+    @ApiOperation(value = "Add a new zenodo.org token, used by accounts page.", authorizations = {
+            @Authorization(value = JWT_SECURITY_DEFINITION_NAME) }, notes = "This is used as part of the OAuth 2 web flow. "
+            + "Once a user has approved permissions for Collaboratory"
+            + "Their browser will load the redirect URI which should resolve here", response = Token.class)
+    public Token addZenodoToken(@ApiParam(hidden = true) @Auth User user, @QueryParam("code") String code) {
+        //if (code.isEmpty()) {
+        //    throw new CustomWebApplicationException("Please provide an access code", HttpStatus.SC_BAD_REQUEST);
+        //}
+
+        final AuthorizationCodeFlow flow = new AuthorizationCodeFlow.Builder(BearerToken.authorizationHeaderAccessMethod(), HTTP_TRANSPORT,
+                JSON_FACTORY, new GenericUrl(ZENODO_URL + "oauth/token"),
+                new ClientParametersAuthentication(zenodoClientID, zenodoClientSecret), zenodoClientID,
+                ZENODO_URL + "oauth/authorize").build();
+
+        LOG.info("About to try and grab zenodo access token");
+        String accessToken;
+        String refreshToken;
+        try {
+            TokenResponse tokenResponse = flow.newTokenRequest(code)
+                    .setRequestInitializer(request -> request.getHeaders().setAccept("application/json")).setGrantType("authorization_code")
+                    .setRedirectUri(zenodoRedirectUri).execute();
+            accessToken = tokenResponse.getAccessToken();
+            refreshToken = tokenResponse.getRefreshToken();
+        } catch (IOException e) {
+            LOG.error("Retrieving zenodo access token was unsuccessful");
+            throw new CustomWebApplicationException("Could not retrieve zenodo token based on code", HttpStatus.SC_BAD_REQUEST);
+        }
+
+
+
+        //String url = ZENODO_URL + "api/user";
+        //Optional<String> asString2 = ResourceUtilities.asString(url, accessToken, client);
+        //String username = getUserName(url, asString2);
+        String username = null;
+
+        if (user != null) {
+            List<Token> tokens = tokenDAO.findZenodoByUserId(user.getId());
+
+            if (tokens.isEmpty()) {
+                Token token = new Token();
+                token.setTokenSource(TokenType.ZENODO_ORG);
+                token.setContent(accessToken);
+                token.setRefreshToken(refreshToken);
+                token.setUserId(user.getId());
+                if (username != null) {
```

Comment:
Looking at the code, the username seems to only come into play if it's a source control token, so I think it's OK that you can't fetch it.

I would, although I wonder what OICR folks think:

1. Set it to an empty value instead of the Dockstore username. If we don't have the username, we don't have it and shouldn't put in a value.
2. I would remove this if/else and just have `token.setUsername("")` with a comment.
3. Remove lines 634-637

Reply:
done